flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/constraints/KeyConstraint.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.api.constraints;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.api.ValidationException;
+import org.apache.flink.table.expressions.FieldReferenceExpression;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A unique key constraint. It can be declared also as a PRIMARY KEY.
+ *
+ * @see ConstraintType
+ */
+@PublicEvolving
+public final class KeyConstraint extends AbstractConstraint {
+	private final List<FieldReferenceExpression> columns;
```

Comment:
I'm not sure whether we should use `FieldReferenceExpression` here. The reasons are following:
1. It's harder to construct
2. The fields `inputIndex` and `fieldIndex` fields from `FieldReferenceExpression` might be inappropriate to exposed here. 

Simply use column name and column type is sufficient here.

Reply:
Personally I would prefer to have the `FieldReferenceExpression` here. Exactly for the `inputIndex` and `fieldIndex info, as it might be helpful for a queries that have two inputs (it might not). In general I think schema should be fixed after every modification. In general `fieldIndex` should always be source of truth for any field references.

Is it that hard though? The `KeyConstraint` class is sort of a second level API. Users should rather not use it. They should go against the `TableSchema` which exposes method `primaryKey(String name, String[] columns)`, which performs the resolution.

This resolution has to be performed anyway at some point.

Reply:
I'm not sure I have the same thought about the responsibility of `TableSchema` right now. In my thought, `TableSchema` is kind of metadata of table, or even a data structure which reflects the whole static information of the original DDL. It should be static and table orient. What you suggested sounds like `TableSchema` is bound to a query?

Reply:
It can be. All `QueryOperation`s and even `Table` objects return their `TableSchema`.

Reply:
For the consistency, I think we should treat `TableSchema` as static information, just as other fields do, like watermark and computed column.

Reply:
I had a longer investigation on here. In a longer run I think we should adjust the `QueryOperation` and `Table` a bit so they do not return a full blown `TableSchema`. They should never care about the unique constraints, watermarks etc.

I checked what information other databases store and they do store information that might need to be resolved (e.g. for primary keys column name and position within the declaration of constraint, for computed columns the expression). Therefore if we assume the TableSchema needs to be resolved I do agree with you we should store only the column names.